x-pack/plugins/alerting/server/saved_objects/migrations.ts

```diff
@@ -869,3 +896,42 @@ function pipeMigrations(...migrations: AlertMigration[]): AlertMigration {
   return (doc: SavedObjectUnsanitizedDoc<RawRule>) =>
     migrations.reduce((migratedDoc, nextMigration) => nextMigration(migratedDoc), doc);
 }
+
+/**
+ * This creates a migration map that applies search source migrations to legacy es query rules
+ */
+const getEsQueryAlertSearchSourceMigrations = (searchSourceMigrations: MigrateFunctionsObject) =>
+  mapValues<MigrateFunctionsObject, MigrateFunction>(
+    searchSourceMigrations,
+    (migrate: MigrateFunction<SerializedSearchSourceFields>): MigrateFunction =>
+      (state) => {
+        const _state = state as unknown as { attributes: RawRule };
+
+        const serializedSearchSource = _state.attributes.params
+          .searchConfiguration as SerializedSearchSourceFields;
+
+        if (!serializedSearchSource) return _state;
```

Comment:
I took a closer look and this seems very unsafe to me.

The `SerializedSearchSourceFields` type is an object with lots of keys, all optional, and in this code we're basically saying that *anything* under the key `searchConfiguration` will be treated as this object and we lie to the compiler telling it that the argument passed to the `migrate` parameter is safe.

My recommendation is that we:

1. Fix the typing. Using `as unknown as` here seems unnecessarily unsafe considering a failed migration causes the entire upgrade to fail and the cluster cant start. Migrations are a critical path.
2. Add a reliable way of validating the `searchConfiguration` param by adding a safe `SerializedSearchSourceFields` deserialiser (using `io-ts` perhaps? or at least an `isSerializedSearchSourceFields` function that ensures it's at least an Object).
3. This is abit of a _nit_, but naming a parameter `migrate` in a huge file full of migrations makes this a bit harder to maintain. Perhaps rename to `migrateSerializedSearchSourceFields` or something like that?

Reply:
Added `isSerializedSearchSourceFields` check, `migrate` parameter renamed.